### PR TITLE
Dynamically choose the python binary at runtime

### LINKED ISSUE #532: Proton 3.7-4 beta - Games don't start anymore
i have switched from proton 3.7.3 to 3.7.4 beta, but all "unsupported" games don't start anymore (Gas guzzlers extreme, call of juarez, PES 2018...) 

The one that starts is Tekken 7 which was in the official list.

Python 2.7 and 3 packages are installed (verified using synaptic) as they were before but nevertheless it's probably something python related. Switching back to proton 3.7.3 games start again.

System 
Steam client version: 25/08/2018
Distribution: Ubuntu 18.04 LTS
Opted into Steam client beta?: Yes
Have you checked for system updates?: Yes
CPU: AMD Ryzen5 1500
RAM: 8GB
Video Card: NVIDIA GTX750ti
Driver version: NVidia 396.54
Video memory: 2GB

### PR DESCRIPTION
**Pull request by MayeulC**
This builds on my comments in #898 and #532. It should also close #532.
@aeikum, @Elkasitu and @nanonyme might be interested.

The code was tested in place of the steam-distributed proton, but only on Arch Linux, trough flatpak. The build script was not tested either.

Tell me what you think. Here are a few more comments:

# Benefits
Such an approach brings:
* broader compatibility (forward, backwards and does not require extra software)
* better error messages (helpfully tell the user that python isn't installed)
* the main executable is POSIX-compliant
* this feels less hacky than to code for platform-specific configuration, and to modify the script with sed.

# Drawbacks:
* Such a workaround is likely to stay for a very long time if merged (unless telemetry tells us it's OK to drop backwards compatibility in a few years)
* No *real* user-facing advantages over the current approach
* Needs a supplementary workaround if `proton` needs to work from the `$PATH`
* *I feel like I'm forgetting one, I'll add it back if I can remember it*

# Notes:
As-is, it prefers the binaries in the following order: `python3`, `python`, `python2`.

### CODE REVIEW

**Comment by MayeulC** [comment]
Btw, @Elkasitu, I attributed one of the commits there to you, even if that's not *exactly* your commit, but rather a part of it. Please tell me if you are not OK with that.

**Comment by nanonyme** [comment]
Maybe the one drawback you're forgetting is this is a runtime check whereas the current thing is a build-time check?

File: proton.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env @PYTHON_NAME@
+#!/usr/bin/env python
```
Comment (Elkasitu) [inline proton.py:1]: is this still necessary given the proton sh script?

**Comment by MayeulC** [comment]
Why would that be a drawback, though? The overhead will be negligible, depending on the needs. I timed it to 0.1 seconds in the worst case on the not-so-powerful laptop I'm typing this on.

It adds a bit of complexity, though, and one more file to the repository.

File: proton [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+pythons="python3 python python2"
+
+# Get proton's path, relative to this script's
+# Note: this won't work if it was called from the $PATH
+PROTON_PATH=$(dirname "$0")/proton.py
+
+for python in $pythons
+do
+	# Try every python. The "x" is here to avoid an empty string
+	if [ $($python -c "print('success')" 2>/dev/null )x = "success"x ]
```
Comment (Elkasitu) [inline proton:13]: why not something like `if [ -x "$(command -v $pybin)" ]` ? The current solution works but seems hacky

**Review by Elkasitu — APPROVED** [review]
LGTM, I don't think the fact that it cannot be run from path is an issue as `proton.py` assumes it is being run from the proton dir, if it ever changes to be run from path, then changing the shell script will be nothing compared to the changes to `proton.py`

File: proton (hunk 2 above)
Comment (MayeulC) [reply proton:13]: I thought about this as well, but this has the nice side effect of ensuring that `$python` really works as intended, at a marginal speed tradeoff. This doesn't really feel hacky to me, but I'm open to discussion.

File: proton.py (hunk 1 above)
Comment (MayeulC) [reply proton.py:1]: I think it's good practice to leave it here, and it allows directly executing the script if you feel so inclined (or if distributions end up packaging `Proton`, they can just skip the shell script and rename this). But yes, it isn't strictly necessary.

**Comment by nanonyme** [comment]
I still feel this is most likely overkill. Python 2 will go end-of-life in little over a year requiring a full re-plan of this entire setup (since macOS might remove Python completely at that point) and the current thing most likely handles most if not all cases where system is reasonably new to work with Proton at all.

File: proton.py [hunk 3]
```diff
@@ -84,7 +84,7 @@ def mergedirs(src, dst):
     log("No compat data path?")
     sys.exit(1)
 
-basedir = os.path.dirname(sys.argv[0])
+basedir = os.path.dirname(os.path.realpath(__file__))
```
Comment (nanonyme) [inline proton.py]: This resolves symlinks. Is that intended?

Comment (MayeulC) [reply proton.py]: No, that was not intended. That's a good question, and I am not sure what the correct behaviour should be, as I cherry-picked that change without giving it further thought. The only problematic case I can see is if the script itself is a symlink. If we want to retain compatibility with that case, it should be `abspath` instead. As I don't see any other problems with `abspath`, I'm going to change this. Thanks!

**Comment by MayeulC** [comment]
Made mostly obsolete by a84120449d8177fa40f3de729e55d238d3c24877

**Pull request closed without merging**